### PR TITLE
PARQUET-1002: Compute statistics based on Sort Order

### PR DESCRIPTION
**Pull request by majetideepak**
@lomereiter,  You might also want to take a look since you previously implemented the Statistics API.

### CODE REVIEW

File: src/parquet/statistics.cc [hunk 1]
```diff
@@ -69,6 +69,7 @@ TypedRowGroupStatistics<DType>::TypedRowGroupStatistics(
   IncrementNullCount(null_count);
   IncrementDistinctCount(distinct_count);
 
+  comparator_ = std::static_pointer_cast<CompareDefault<DType> >(Compare::Make(schema));
```
Comment (lomereiter) [inline src/parquet/statistics.cc]: Should this line be just moved into `SetDescr` then?

File: src/parquet/util/comparison.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+#include "parquet/exception.h"
+#include "parquet/schema.h"
+#include "parquet/types.h"
+#include "parquet/util/comparison.h"
+
+namespace parquet {
+
+std::shared_ptr<Compare> Compare::Make(const ColumnDescriptor* descr) {
+  if (SortOrder::SIGNED == GetSortOrder(descr->logical_type(), descr->physical_type())) {
```
Comment (lomereiter) [inline src/parquet/util/comparison.cc]: I counted 5 occurrences of this call, it probably makes sense to add a convenience method to `ColumnDescriptor`

Comment (majetideepak) [inline src/parquet/util/comparison.cc]: Yes! makes sense.

**Comment by majetideepak** [comment]
@wesm, @xhochy can you take a look at this? thanks!

**Comment by wesm** [comment]
I'm sorry about the delay, I will try to review this today or worst case tomorrow

**Comment by wesm** [comment]
Going to review this today

File: src/parquet/types.h [hunk 3]
```diff
@@ -283,6 +296,11 @@ PARQUET_EXPORT std::string FormatStatValue(Type::type parquet_type, const char*
 
 PARQUET_EXPORT int GetTypeByteSize(Type::type t);
 
+SortOrder::type PARQUET_EXPORT DefaultSortOrder(Type::type primitive);
+
+SortOrder::type PARQUET_EXPORT GetSortOrder(LogicalType::type converted,
```
Comment (wesm) [inline src/parquet/types.h]: Can you put the `PARQUET_EXPORT` before the return type for consistency (we should go through and make them all consistent)?

File: src/parquet/util/comparison.h [hunk 4]
```diff
@@ -20,40 +20,145 @@
...
 #include "parquet/types.h"
 
 namespace parquet {
 
-template <typename T>
-struct Compare {
-  explicit Compare(const ColumnDescriptor* descr) : type_length_(descr->type_length()) {}
-
-  inline bool operator()(const T& a, const T& b) { return a < b; }
+class PARQUET_EXPORT Compare {
```
Comment (wesm) [inline src/parquet/util/comparison.h]: nit: can we rename this `Comparator`?

File: src/parquet/util/comparison.h [hunk 5]
```diff
@@ -20,40 +20,145 @@
...
- private:
-  int32_t type_length_;
+// The default comparison is SIGNED
+template <typename DType>
+class PARQUET_EXPORT CompareDefault : public Compare {
+ public:
+  typedef typename DType::c_type T;
+  CompareDefault() {}
+  virtual ~CompareDefault() {}
+  virtual bool operator()(const T& a, const T& b) { return a < b; }
```
Comment (wesm) [inline src/parquet/util/comparison.h]: override (instead of virtual)

File: src/parquet/util/comparison.h [hunk 6]
```diff
@@ -20,40 +20,145 @@
...
 
 template <>
-inline bool Compare<Int96>::operator()(const Int96& a, const Int96& b) {
-  return std::lexicographical_compare(a.value, a.value + 3, b.value, b.value + 3);
-}
+class PARQUET_EXPORT CompareDefault<Int96Type> : public Compare {
+ public:
+  CompareDefault() {}
+  virtual ~CompareDefault() {}
+  virtual bool operator()(const Int96& a, const Int96& b) {
```
Comment (wesm) [inline src/parquet/util/comparison.h]: override

File: src/parquet/util/comparison.h [hunk 7]
```diff
@@ -20,40 +20,145 @@
...
-inline bool Compare<ByteArray>::operator()(const ByteArray& a, const ByteArray& b) {
-  auto aptr = reinterpret_cast<const int8_t*>(a.ptr);
-  auto bptr = reinterpret_cast<const int8_t*>(b.ptr);
-  return std::lexicographical_compare(aptr, aptr + a.len, bptr, bptr + b.len);
-}
+class PARQUET_EXPORT CompareDefault<ByteArrayType> : public Compare {
+ public:
+  CompareDefault() {}
+  virtual ~CompareDefault() {}
+  virtual bool operator()(const ByteArray& a, const ByteArray& b) {
```
Comment (wesm) [inline src/parquet/util/comparison.h]: override

File: src/parquet/util/comparison.h [hunk 8]
```diff
@@ -20,40 +20,145 @@
...
-  auto aptr = reinterpret_cast<const int8_t*>(a.ptr);
-  auto bptr = reinterpret_cast<const int8_t*>(b.ptr);
-  return std::lexicographical_compare(aptr, aptr + type_length_, bptr,
-                                      bptr + type_length_);
-}
+class PARQUET_EXPORT CompareDefault<FLBAType> : public Compare {
+ public:
+  explicit CompareDefault(int length) : type_length_(length) {}
+  virtual ~CompareDefault() {}
+  virtual bool operator()(const FLBA& a, const FLBA& b) {
```
Comment (wesm) [inline src/parquet/util/comparison.h]: override

File: src/parquet/file/metadata.cc [hunk 9]
```diff
@@ -35,62 +35,20 @@ const ApplicationVersion ApplicationVersion::PARQUET_251_FIXED_VERSION =
...
-    case LogicalType::MAP:
-    case LogicalType::MAP_KEY_VALUE:
-    case LogicalType::INTERVAL:
-    case LogicalType::NONE:  // required instead of default
-      return SortOrder::UNKNOWN;
-  }
-  return SortOrder::UNKNOWN;
-}
+const ApplicationVersion ApplicationVersion::PARQUET_CPP_FIXED_STATS_VERSION =
+    ApplicationVersion("parquet-cpp version 1.2.1");
```
Comment (majetideepak) [inline src/parquet/file/metadata.cc]: Releasing 1.2.1 sounds like a good idea. But this value should work even if we make a 1.3.0 release.

File: src/parquet/file/metadata.cc [hunk 10]
```diff
@@ -577,16 +539,24 @@ class ColumnChunkMetaDataBuilder::ColumnChunkMetaDataBuilderImpl {
...
+    stats.min_value = val.min();
+    stats.__isset.min_value = val.has_min;
+    stats.__isset.max_value = val.has_max;
     stats.__isset.null_count = val.has_null_count;
     stats.__isset.distinct_count = val.has_distinct_count;
+    // If the order is SIGNED, then the old min/max values must be set too.
+    // This for backward compatibility
+    if (is_signed) {
+      stats.max = val.max();
+      stats.min = val.min();
```
Comment (majetideepak) [inline src/parquet/file/metadata.cc]: I check and it didn't.

File: src/parquet/parquet_reader_writer-test.cc [hunk 11]
```diff
@@ -0,0 +1,330 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (majetideepak) [inline src/parquet/parquet_reader_writer-test.cc]: My motivation is to add end to end writer-reader tests for Parquet which we currently don't have.

File: src/parquet/parquet_reader_writer-test.cc [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+  stats_[0]
+      .set_min(std::string(reinterpret_cast<const char*>(&values_[0]), sizeof(T)))
+      .set_max(std::string(reinterpret_cast<const char*>(&values_[9]), sizeof(T)));
+}
+
+// TYPE::DOUBLE
+template <>
+void TestStatistics<DoubleType>::AddNodes(std::string name) {
+  // Double physical type has only Signed Statistics
+  fields_.push_back(schema::PrimitiveNode::Make(name, Repetition::REQUIRED, Type::DOUBLE,
```
Comment (majetideepak) [inline src/parquet/parquet_reader_writer-test.cc]: I will take a look.

File: src/parquet/parquet_version.h [hunk 13]
```diff
@@ -19,6 +19,6 @@
 #define PARQUET_VERSION_H
 
 // define the parquet created by version
-#define CREATED_BY_VERSION "parquet-cpp version 1.1.1-SNAPSHOT"
+#define CREATED_BY_VERSION "parquet-cpp version 1.2.1-SNAPSHOT"
 
-#endif  // PARQUET_VERSION_H
+#endif // PARQUET_VERSION_H
```
Comment (majetideepak) [inline src/parquet/parquet_version.h]: Removing this file causes one of the Travis CI tests to fail. I am not sure why. I will look into this as well.

File: src/parquet/types.h (hunk 3 above)
Comment (majetideepak) [inline src/parquet/types.h]: Will do.

File: src/parquet/util/comparison.h (hunk 4 above)
Comment (majetideepak) [inline src/parquet/util/comparison.h]: Sure!

File: src/parquet/util/comparison.h (hunk 5 above)
Comment (majetideepak) [inline src/parquet/util/comparison.h]: will do here and below.

File: src/parquet/file/metadata.cc (hunk 9 above)
Comment (majetideepak) [inline src/parquet/file/metadata.cc]: I just noticed https://github.com/apache/parquet-format/pull/55/files adds some more changes to the min-max stats format. I think we should wait until that gets merged.

**Comment by majetideepak** [comment]
Thanks for the feedback @wesm! As far as I remember, the sync discussions were geared more towards UTF-8  (and other string encodings) and not much on Int96. Int96 is deprecated and is currently used only for Impala Timestamp values which require a custom sort order.

I have also implemented this patch with the idea of users be able to implement custom `Comparator` classes and plugin into the writer properties. Any custom order should be easy to implement now by simply extending the `CompareDefault**Type**` class

File: src/parquet/column_writer.cc [hunk 14]
```diff
@@ -267,7 +267,9 @@ int64_t ColumnWriter::Close() {
     FlushBufferedDataPages();
 
     EncodedStatistics chunk_statistics = GetChunkStatistics();
-    if (chunk_statistics.is_set()) metadata_->SetStatistics(chunk_statistics);
+    if (chunk_statistics.is_set())
```
Comment (xhochy) [inline src/parquet/column_writer.cc]: Nit: Please use braces.

File: src/parquet/file/metadata.h [hunk 15]
```diff
@@ -92,7 +82,7 @@ class ApplicationVersion {
   bool VersionEq(const ApplicationVersion& other_version) const;
 
   // Checks if the Version has the correct statistics for a given column
-  bool HasCorrectStatistics(Type::type primitive) const;
+  bool HasCorrectStatistics(Type::type primitive, SortOrder::type sort_order) const;
```
Comment (xhochy) [inline src/parquet/file/metadata.h]: As we change the interface here, we should make the new version 1.3.0. 

@majetideepak alternatively can we provide a default argument here?

**Comment by wesm** [comment]
@majetideepak if you can incorporate feedback and make this not WIP, we should be able to cut the 1.3.0 RC after this is in

**Comment by majetideepak** [comment]
@wesm Sorry for the delay! I somehow missed your comment. I will incorporate this feedback tomorrow morning.

File: src/parquet/parquet_version.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (majetideepak) [reply src/parquet/parquet_version.h]: Filed VER-1088

File: src/parquet/util/comparison.h (hunk 5 above)
Comment (majetideepak) [reply src/parquet/util/comparison.h]: `CompareDefault` is the base class and will require virtual. I used override inside the derived classes.

**Comment by wesm** [comment]
A follow up question on the comparator thing. Any way to do this as fully compile-time with no virtual functions (and so no exported symbols in the DLL)?

**Comment by majetideepak** [comment]
I couldn't come up with an elegant design to make it fully-compile time.
The comparator depends upon the `SortOrder` and the `c_type` of the parquet physical types.
The `c_type` can be determined at compile time for a given TypedRowGroupStatistics. I don't know if we can determine the `SortOrder` at compile time.
Another motivation for virtual functions is to allow users to specify custom comparators for unknown sort orders.

File: src/parquet/parquet_reader_writer-test.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (xhochy) [reply src/parquet/parquet_reader_writer-test.cc]: I would also like to see this in `statistics-test.cc`. We can move it once we have full end-to-end tests.

**Comment by majetideepak** [comment]
Fixed issue https://github.com/apache/parquet-cpp/pull/383/files#r134647005
We can go ahead without the merge of apache/parquet-format#55. That PR only impacts how the SortOrder is inferred. This PR will not require any changes.

**Comment by xhochy** [comment]
Travis is failing with 

```
/Users/travis/build/apache/parquet-cpp/src/parquet/statistics-test.cc:634:13: error: variable-sized object may not be initialized
  char vals[NUM_VALUES][FLBA_LENGTH] = {"b12345", "a12345", "c12345", "d12345", "e12345",
```

Once that is fixed, feel free to merge.

**Comment by majetideepak** [comment]
@xhochy Fixed the error. I don't think I have merge access yet!

**Comment by wesm** [comment]
Merging. It looks like you haven't been added to the LDAP group, @julienledem can you take a look?

**Pull request closed without merging**